src/full_client.toit

```diff
@@ -26,37 +69,98 @@ The first side (reading or writing) that detects an issue with the transport dis
 Also sends ping requests to keep the connection alive (if $keep_alive is called).
 */
 class Connection_:
+  /**
+  The connection is considered alive.
+  This could be because we haven't tried to establish a connection, but it
+    could also be that we are happily running.
+  */
+  static STATE_ALIVE_ ::= 0
+  /**
+  The connection is closed or in the process of closing.
+  Once connected, if there is an error during receiving or sending, the connection will
+    switch to this state and call $Transport.close. This will cause the
+    other side (receive or send) to shut down as well (if there is any).
+  */
+  static STATE_CLOSED_ ::= 1
+
+  state_ / int := STATE_ALIVE_
+
   transport_ / ActivityMonitoringTransport
+  reader_ /reader.BufferedReader
+  writer_ /writer.Writer
+  writing_ /monitor.Mutex ::= monitor.Mutex
+  is_writing_ /bool := false
+
+  closing_reason_ /any := null
+
+  keep_alive_duration_ /Duration?
+  activity_task_ /Task_? := null
 
   /** Constructs a new connection. */
   constructor .transport_ --keep_alive/Duration?:
-    throw "UNIMPLEMENTED"
+    reader_ = reader.BufferedReader transport_
+    writer_ = writer.Writer transport_
+    keep_alive_duration_ = keep_alive
 
-  is_alive -> bool: throw "UNIMPLEMENTED"
-  is_closed -> bool: throw "UNIMPLEMENTED"
+
+  is_alive -> bool: return state_ == STATE_ALIVE_
+  is_closed -> bool: return state_ == STATE_CLOSED_
 
   /**
   Starts a task to keep this connection to the broker alive.
   Sends ping requests when necessary.
   */
   keep_alive --background/bool:
-    throw "UNIMPLEMENTED"
+    assert: background
+    if activity_task_: throw "ALREADY_RUNNING"
+    if keep_alive_duration_ == (Duration --s=0): return
```

Comment:
It would be nice with a better zero check for durations.

Reply:
switched to `Duration.ZERO` but will also add an `is_zero` method to `Duration`.